### Stats with Python

- Centrality Measures
- Measures of Dispersion
- Random Numbers & Sampling
- Probability Distributions
- Random Distributions
- Hypothesis Testing
- Statistical Models
    - Linear Models
    - Non-linear Models
    - Design Matrices with **patsy**
    - Statistical Modeling with **statsmodels**
- Discrete Regression Models
    - Logit
    - MNLogit
    - Poission
- ANOVA Test of Fitted Regression Model

In [1]:
import numpy as np
from scipy import stats

**Centrality Measures**
- mean, median, mode

In [2]:
s1 = np.array([86, 47, 45, 47, 40])
print(np.mean(s1))
print(np.median(s1))
# If there are more than a single value, then only the first one is returned by mode function.
print(stats.mode(s1))
print(stats.mode(s1)[0][0])

53.0
47.0
ModeResult(mode=array([47]), count=array([2]))
47


**Measures of Dispersion** - These measures provide insights on the spread of a given dataset.

Major measures of dispersion are: 
- range, percentile, inter-quartile range, standard deviation, variance, skewness, and kurtosis.


In [3]:
print("Range: {}".format(np.ptp(s1)))
s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])
# 45th percentile refers to a value below which 45% of data points are found.
print("Percentile: {}".format(np.percentile(s2, 45, interpolation='lower')))
print("Quartiles: {}".format(np.percentile(s2, [25, 50, 75], interpolation='lower')))
print("IQR: {}".format(stats.iqr(s2, rng=(25, 75), interpolation='lower')))
print("Variance: {}".format(np.var(s2)))
print("Standard Deviation: {}".format(np.std(s2)))
# s2 representing a population's sample (i.e. sample derived from a population)
print("Sample Variance: {}".format(np.var(s2, ddof=1)))
print("Sample Standard Deviation: {}".format(np.std(s2, ddof=1)))
# Positive: right-skewed; Negative: left-skewed; Zero: Unskewed
print("Skewness: {}".format(stats.skew(s2)))
# Kurtosis indicates how much of data is concentrated around mean or shape of the probability distribution.
# Default: Fisher’s definition. Change to Pearson by setting fisher parameter to False.
print("Kurtosis: {}".format(stats.kurtosis(s2)))

Range: 46
Percentile: 65
Quartiles: [47 65 83]
IQR: 36
Variance: 454.21000000000004
Standard Deviation: 21.312203077110542
Sample Variance: 504.6777777777778
Sample Standard Deviation: 22.46503455990615
Skewness: 0.04321019390325423
Kurtosis: -1.5694354898634155


**Random Numbers & Sampling**

In [4]:
# A random number is a number, chosen by chance from a distribution.
# rand function generates uniformly distributed numbers from range [0, 1)
print(np.random.rand())
# generates a 2*3 array
print(np.random.rand(2,3))
# Select two samples w/o replacement
print(np.random.choice([11, 22, 33], 2, replace=False))
# seed is a number that sets the initial state of random number generator.
np.random.seed(100)
print(np.random.rand())
np.random.seed(100)
print(np.random.rand())

0.4882053338415546
[[0.87317956 0.2490897  0.76373767]
 [0.10786928 0.45116403 0.54780722]]
[11 22]
0.5434049417909654
0.5434049417909654


### Probability Distributions

There are two types of probability distributions namely **discrete** and **continuous** that take **integer** and **real** values, respectively.

**scipy.stats** module provides classes that represent random variables, corresponding to a large number of probability distributions.

**Example:** the class **norm** represents normal continuous random variable, and **binom** represents binomial discrete random variable.

### Random Distributions

**scipy.stats** module provide a lot of methods for created discrete and continuous random variables.

Commonly used methods are described below.

    pdf/ pmf: Probability distribution function (continuous) or probability mass function (discrete).
    cdf: Cumulative distribution function.
    sf: Survival function (1 – cdf).
    rvs: Creating random samples from a defined distribution.



In [5]:
# A normal continuous random variable of mean 1.0 and std 2.5.
x = stats.norm(loc=1.0, scale=2.5)
print(x)
# Estimates probabilities and cumulative probabilities at -1, 0 and 1.
print(x.pdf([-1, 0, 1]))
print(x.cdf([-1, 0, 1]))
# Generates six random numbers from defined normal distribution.
print("Random Numbers: \n {}".format(x.rvs((2,3))))
# Sample mean
print("Sample Mean: {}".format(np.mean(x.rvs((2,3)))))
# Distribution mean
print("Distribution Mean: {}".format(x.mean()))

[0.11587662 0.14730806 0.15957691]
[0.2118554  0.34457826 0.5       ]
Random Numbers: 
 [[-0.40409689 -3.12269376  1.88668613]
 [-0.96516083  0.42031951  1.5199392 ]]
Sample Mean: 0.7768877525594577
Distribution Mean: 1.0


**Problem**

Simulate a random experiment of tossing a coin 10,000 times, and determine the count of Heads returned?

In [6]:
np.random.seed(1)
re = stats.binom(n=1, p=0.5)
binom_vars = re.rvs(size=10000)
print(binom_vars)
print(np.bincount(binom_vars))

[0 1 0 ... 0 0 1]
[4990 5010]


### Hypothesis Testing

 - Define the null hypothesis and the alternative hypothesis.
 - Select a test statistics whose probability distribution function can be found under the null hypothesis.
 - Collect data.
 - Compute the test statistics from the data and calculate its p-value under the null hypothesis.
 - Null hypothesis is rejected if the p-value is lower than predetermined significance value.
 
In hypothesis testing, selecting a test statistics is the most difficult part.

The methods used for performing **t-test** are shown below:

    stats.ttest_1samp: Tests if the mean of a population is a given value.
    stats.ttest_ind: Tests if the means of two independent samples are equal.
    stats.ttest_rel: Tests if the means of two paired samples are equal.
    
Let's consider a common hypothesis: Mean of a population is equal to certain value.

In reality, we estimate mean and variance of a sample and calculate the test statistic.
- If population variance is identified, then it is reasonable to consider that test statistic is normally distributed.
- If population variance is unknown, then sample variance is used, and test statistic follows t-distribution.
    
    

In [7]:
# Consider a normal population with mean 0.8 and standard deviation 0.5.
# Define the null hypothesis as Mean of the population is 1.0.
# Let's calculate t-statistic and p-value.

mu, sigma = 0.8, 0.5
X = stats.norm(mu, sigma)

# Deriving a sample
n = 100
X_sample = X.rvs(n)

# Computing test statistic
t, p = stats.ttest_1samp(X_sample, 1.0)
print(t, p)

# Conclusion: Since p-value is very low and less than the significance level 0.05, 
# we can reject the null hypothesis and infer that the mean of the population is not 1.


-2.9673196734975202 0.0037666655277120245


In [8]:
# Let's consider another problem, where the null hypothesis states that the population means of two random variables are equal.

# Get two samples from different populations
X1 = stats.norm(0.25, 1.0)
X2 = stats.norm(0.50, 1.0)

X1_sample = X1.rvs(100)
X2_sample = X2.rvs(100)

t, p = stats.ttest_ind(X1_sample, X2_sample)
print(t, p)

# Conclusion: Since p-value is greater than the significance level 0.05, we can accept the null hypothesis and state that population means of both samples are equal.


-2.6036006833310372 0.00992383762438619


**Examples**

In [9]:
import numpy as np
from scipy import stats
s1 = [45, 38, 52, 48, 25, 39, 51, 46, 55, 46]
s2 = [34, 22, 15, 27, 37, 41, 24, 19, 26, 36]
t, p = stats.ttest_ind(s1, s2)
print(t, p)

4.257546665558161 0.0004736633119019225


In [10]:
import numpy as np
from scipy import stats
s1 = [12, 7, 3, 11, 8, 5, 14, 7, 9, 10]
s2 = [8, 7, 4, 14, 6, 7, 12, 5, 5, 8]
t, p = stats.ttest_rel(s1, s2)
print(t, p)

1.315587028960544 0.2208380130273219


### Statistical Model

A statistical model is a mathematical equation, which explains the relationship between dependent variables (Y) and independent variables (X).

In general, a model is written as

$$ Y = f(X) $$

However, in reality, an element of uncertainty is expected due to factors such as measurement noise. Hence the aforementioned equation can be rewritten as

$$ Y = f(X) + e $$ 

where 'e' is residual error.
    
#### Statistical Modeling
    
Statistical Modeling deals with creating models that attempt to explain the data best.

#### Linear Models

The simplest model is a **linear** model represented as $$ Y = B_0 + B_1 * X + e $$ where the coefficients, B0 and B1 are the parameters of the model and e is normally distributed residual error.

 - A linear regression model assumes that residuals are independent and normally distributed.
 - The model is fitted to data using ordinary least squares approach.
 
In most of the linear regression models, a dependent variable Y is written as:

    a linear combination of the response variables X i.e 
$$ Y = B_0 + B_1 * X_1 + ... + B_n * X_n $$, or
    
    functions of the response variables i.e 

$$ Y = B_0 + B_1 * X^1 + B_2 * X^2 + ... + B_n * X^n $$, or
    
    models that have a linear component i.e 
    
$$ Y = B_0 + B_1 * sin(X_1) + B_2 * cos(X_2) $$
    
#### Non-Linear Models

Other than linear regression models, statistical modeling can be used to build non-linear models.

- Errors of dependent variable follow a distribution other than normal distribution.
- Examples of non-linear models are: **Binomial Regression** and **Poisson Regression**.
- In most of the cases, the non-linear models are generalized to linear models.

#### Design Matrices

In reality, we choose a model and fit the available data to it.

Once a model is chosen, design matrices y and X are constructed, and the regression problem is written, in matrix form, as: $$ y = XB + e $$

    where y is the vector of dependent variable, X is the vector of observations, B is a vector of coefficients, and e is the residual (error).

Thus the obtained design matrices are passed as inputs to the chosen model.

#### Statistical Modelling with StatsModels

- statsmodels library supports several types of statistical models.
- All of them follow a similar usage pattern.
- A statistical model is represented by a model class.
- A model can be initiated with,
    - given design matrices of dependent and independent variables, or with
    - given patsy formula and data frame or dictionary-like object.
    
Step 1: Creating a Model

An instance of a model class is created in either of the following ways.

    model = sm.MODEL(y, X) or
    model = smf.model(patsy_formula, data)

    where MODEL and model refer to model names such as OLS, GLM, ols, glm, etc. Uppercase names take design matrices as arguments, and lowercase names take Patsy formulas and data frames as arguments.
    
**Note**: We need to add bias term (constant) manually while using **statsmodels.api** as **sm** whereas the constant is automatically added with **statsmodels.formula.api** as **smf**.

Step 2: Fitting a Model

In order to fit the model with data, fit method is invoked on the created model, as shown below.

    result = model.fit()

The fit method returns a result object, which has methods and attributes for further analysis.

Step 3: Viewing Model Summary

The summary method of result object produces a summary text that describes the result of the fit, as shown below.

    print(result.summary())
    
The displayed summary text varies for each statistical model and provides information of various statistical parameters.

Step 4: Analyzing the Model

Other than viewing the summary statistics, we can perform operations like,

    Determining fitted values,
    Predicting the dependent variable values for new independent variable values.
    Checking if residuals of fitted models follow a normal distribution or not.


In [11]:
y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
X = np.vstack([np.ones(5), x1, x2, x1*x2]).T
print(y)
print(X)

[1 2 3 4 5]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


#### Design Matrices with patsy

- patsy, a python library, allows defining a model easily.
- It also constructs relevant design matrices, automatically, using **patsy.dmatrices** function.
- patsy.dmatrices takes a formula (in string form) as a first argument, and a dictionary-like object with data arrays for the response variables as a second argument.
    
    **Formulae**
    
    'y ~ x' : y is linearly dependent on x. ~ symbol separates dependent variable from independent variable terms. This is also equivalent to 'y ~ 1 + x'.

    'y ~ x1 + x2' : y is a linear combination of x1 and x2. + sign is used to denote the union of terms.

    'y ~ x1 * x2' : x1 * x2 is an interaction term that includes all lower order terms.
     
    'y ~ np.log(x1)': Often numpy functions can be used to transform terms in the expression.

    'y ~ I(x1 + x2)': I is the identify function, used to escape arithmetic expressions and are evaluated.

    'y ~ C(x1)': Treats the variable x1 as a categorical variable.

In [12]:
import patsy
y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
data = {'y':y, 'x1':x1, 'x2':x2}

y, X = patsy.dmatrices('y ~ 1 + x1 + x2 + x1*x2', data)

print(y)
print(X)

[[1.]
 [2.]
 [3.]
 [4.]
 [5.]]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


#### Example Datasets

statsmodels contain few popular example datasets, which can be used to explore various utilities of the package.

The example data sets are available in the datasets module.

Each dataset is associated with special variables like SOURCE, DESCSHORT, DESCLONG, that provide more informationa about the dataset.

A dataset can be loaded using load function, and its data can be accessed using data attribute in the form of Numpy's recarray.

In [13]:
import statsmodels.api as sm
bc_cancer_set = sm.datasets.cancer
bc_cancer = bc_cancer_set.load()
bc_cancer_data = bc_cancer.data
print(type(bc_cancer_data))

<class 'numpy.recarray'>


In [14]:
# Alternative - load pandas
bc_cancer_df = bc_cancer_set.load_pandas()
type(bc_cancer_df)

statsmodels.datasets.utils.Dataset

### Modeling Experiments

In [15]:
import statsmodels.formula.api as smf
icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")
icecream_data = icecream.data
print(icecream_data.columns)

linear_model1 = smf.ols('cons ~ price + temp', icecream_data)
linear_result1 = linear_model1.fit()
print(linear_result1.summary())

Index(['cons', 'income', 'price', 'temp'], dtype='object')
                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.633
Model:                            OLS   Adj. R-squared:                  0.606
Method:                 Least Squares   F-statistic:                     23.27
Date:                Tue, 28 Jul 2020   Prob (F-statistic):           1.34e-06
Time:                        16:57:12   Log-Likelihood:                 54.607
No. Observations:                  30   AIC:                            -103.2
Df Residuals:                      27   BIC:                            -99.01
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------

#### Discrete Regression

A discrete dependent variable takes few possible outcome values and is not normally distributed.

Hence, linear regression cannot be applied to a discrete variable.

#### Discrete Regression models

statsmodels provide the following classes to work with discrete regression problems.

    Logit: for Logistic Regression
    MNLogit: for Multinomial Logistic Regression
    Poisson: for Poisson Regression
    
Error or outcome values are not normally distributed.


In [16]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd

iris = sm.datasets.get_rdataset("iris").data 
print(iris.info())
print(iris.Species.unique())

iris_subset = iris[(iris.Species == "versicolor") | (iris.Species == "virginica")].copy()
print(iris_subset.Species.unique())

iris_subset.Species = iris_subset.Species.map({"versicolor": 1, "virginica": 0}) 
iris_subset.rename(columns={"Sepal.Length": "Sepal_Length", "Sepal.Width": "Sepal_Width", "Petal.Length": "Petal_Length", "Petal.Width": "Petal_Width"}, inplace=True) 
model = smf.logit("Species ~ Petal_Length + Petal_Width", data=iris_subset)
result = model.fit() 
print("Result Summary:\n", result.summary())

df_new = pd.DataFrame({"Petal_Length": np.random.randn(20) * 0.5 + 5,
                       "Petal_Width": np.random.randn(20) * 0.5 + 1.7})
df_new["P-Species"] = result.predict(df_new)
print(df_new["P-Species"].head(3))

df_new["Species"] = (df_new["P-Species"] > 0.5).astype(int)
df_new.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  150 non-null    float64
 1   Sepal.Width   150 non-null    float64
 2   Petal.Length  150 non-null    float64
 3   Petal.Width   150 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB
None
['setosa' 'versicolor' 'virginica']
['versicolor' 'virginica']
Optimization terminated successfully.
         Current function value: 0.102818
         Iterations 10
Result Summary:
                            Logit Regression Results                           
Dep. Variable:                Species   No. Observations:                  100
Model:                          Logit   Df Residuals:                       97
Method:                           MLE   Df Model:                            2
Date:                Tue, 28 Jul 2020  

Petal_Length  Petal_Width  P-Species  Species
0      3.811639     0.947028   0.999999        1
1      5.695176     2.015813   0.000192        0
2      5.245047     0.695395   0.999601        1
3      4.680638     1.469519   0.951930        1
4      4.669812     1.611719   0.826729        1

- Poisson Model describes a process where dependent variable refers to success count of many attempts and each attempt has a very low probability of success.

In [17]:
awards_df = pd.read_csv("https://stats.idre.ucla.edu/stat/data/poisson_sim.csv")
print(awards_df.head(3))

    id  num_awards  prog  math
0   45           0     3    41
1  108           0     1    41
2   15           0     3    44


There are three type of programs enrolled by a student : 1 - "General", 2 - "Academic", 3 - "Vocational"

In [18]:
poisson_model = smf.poisson('num_awards ~ math + C(prog)', awards_df)
poisson_model_result = poisson_model.fit()
poisson_model_result.summary()

Optimization terminated successfully.
         Current function value: 0.913761
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results                          
==============================================================================
Dep. Variable:             num_awards   No. Observations:                  200
Model:                        Poisson   Df Residuals:                      196
Method:                           MLE   Df Model:                            3
Date:                Tue, 28 Jul 2020   Pseudo R-squ.:                  0.2118
Time:                        16:57:13   Log-Likelihood:                -182.75
converged:                       True   LL-Null:                       -231.86
Covariance Type:            nonrobust   LLR p-value:                 3.747e-21
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -5.2471      0.658     -7.969      0.000      -6.538      -3.957
C(prog)[T.2]     1.0839      0.358      3.025      0.002       0.382       1.786
C(prog)[T.3]     0.3698      0.441      0.838      0.402      -0.495       1.234
math             0.0702      0.011      6.619      0.000       0.049       0.091
================================================================================
"""

The coefficient for math variable is 0.07, which means for every one unit increase in math, the log count increases by 0.07.

Having enrolled for prog=2, i.e., "Academic", instead of "Generic" program, changes the log count by 1.08.

Having enrolled for prog=3, i.e., "Vocational", instead of "Generic" program, changes the log count by 0.37.

In [19]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd
insurance_set = sm.datasets.get_rdataset('Insurance','MASS')
insurance_set_data = insurance_set.data
insurance_model = smf.poisson('Claims ~ np.log(Holders)', insurance_set_data)
insurance_model_result = insurance_model.fit()
print(insurance_model_result.resid)

Optimization terminated successfully.
         Current function value: 3.468160
         Iterations 7
0      5.224177
1     -6.990384
2    -19.554411
3    -45.058842
4     18.332705
        ...    
59    10.466602
60    -0.949599
61     2.084412
62     2.287499
63    12.369671
Length: 64, dtype: float64


#### ANOVA Test

In [20]:
#### Perform ANOVA of a fitted regression model
from statsmodels.stats import anova
icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")
icecream_data = icecream.data
model1 = smf.ols('cons ~ temp', icecream_data).fit()
# let's frame a null hypothesis: Value of B1, i.e., the coefficient of temp variable is zero.

print("ANOVA Test for Model 1:\n", anova.anova_lm(model1))
# The obtained F-statistic is 42.27 and has a very low probability.
# Hence, the null hypothesis can be rejected. i.e., B1 is not equal to zero.

# Now let's create a new model with two independent variables and one response variable.
# With more that one independent variable, the null hypothesis is stated as Coefficients of all independent variables are zero. i.e B1 = 0, and B2 = 0.
# The alternative hypothesis will be that atleast one of the parameters Bj != 0 where j takes the values 1, 2, ...

model2 = smf.ols('cons ~ income + temp', icecream_data).fit()
print("\nANOVA Test for Model 2:\n", anova.anova_lm(model2))

# F-statistic is defined as (Mean square of model/ Mean square of residuals).
# Mean square of model is (sum of squares values of all variables/ degrees of freedom) i.e 
# (0.000288 + 0.087836)/2 = 0.044062.
# Hence, the F-statistic of model is (0.044062 / 0.001385) = 31.813.
# The probability of obtained F-statistic is 7.96*e-08 and is computed as shown below:

print("\nF-statistic: ", stats.f.sf(31.81, 2, 27))


ANOVA Test for Model 1:
             df    sum_sq   mean_sq         F        PR(>F)
temp       1.0  0.075514  0.075514  42.27997  4.789215e-07
Residual  28.0  0.050009  0.001786       NaN           NaN

ANOVA Test for Model 2:
             df    sum_sq   mean_sq          F        PR(>F)
income     1.0  0.000288  0.000288   0.208231  6.518069e-01
temp       1.0  0.087836  0.087836  63.413711  1.470071e-08
Residual  27.0  0.037399  0.001385        NaN           NaN

F-statistic:  7.959504548627583e-08


In [21]:
# ANOVA can also be used to compare two nested or related models.
# model1 regresses consumption with temperature and model2 with temperature and income.
# Below code verifies if the decrease in residuals sum of squares is significant or not.

print(anova.anova_lm(model1, model2))
# p-value i.e 0.005 from the table suggests the decrease in residual sum of squares in model2, compared to model1 is significant.

   df_resid       ssr  df_diff   ss_diff         F    Pr(>F)
0      28.0  0.050009      0.0       NaN       NaN       NaN
1      27.0  0.037399      1.0  0.012611  9.104375  0.005506


/Users/smithkakar/anaconda3/envs/ds/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py:1932: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= _a)


#### Notes
    
    The chi-square test is one of the most commonly used tests to do the goodness of fit tests and is used for discrete distributions like the binomial distribution and the Poisson distribution, whereas The Kolmogorov-Smirnov and Anderson-Darling goodness of fit tests are used for continuous distributions.
    
    Which function is used to test non-correlation between two variables? 
        - scipy.stats.pearsonr(x, y)